lib/updatehub/updatehub.c

```diff
@@ -0,0 +1,799 @@
+		return -1;
+	}
+
+	if (tc_sha256_update(&sha256sum, metadata, strlen(metadata)) == 0) {
+		return -1;
+	}
+
+	if (tc_sha256_final(hash, &sha256sum) == 0) {
+		return -1;
+	}
+
+	memset(_update_info.package_uid, 0, SHA256SUM_STRING_SIZE);
+	for (int i = 0; i < TC_SHA256_DIGEST_SIZE; i++) {
+		snprintk(buffer, TC_SHA256_DIGEST_SIZE, "%02x",
+			 hash[i]);
+		strncat(&_update_info.package_uid[i], buffer, strlen(buffer));
+	}
+
+	return 0;
+}
+
+static bool
+is_compatible_hardware(struct resp_probe_some_boards *metadata_some_boards)
+{
+	int i;
+
+	for (i = 0; i < metadata_some_boards->supported_hardware_len; i++) {
+		if (strcmp(metadata_some_boards->supported_hardware[i],
+			   CONFIG_BOARD) == 0) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static void dns_cb(enum dns_resolve_status status, struct dns_addrinfo *info,
+		   void *user_ip)
+{
+	struct sockaddr_in *addr;
+	char hr_addr[NET_IPV6_ADDR_LEN], *hr_family, *buffer;
+
+	switch (status) {
+	case DNS_EAI_CANCELED:
+		LOG_ERR("DNS query was canceled");
+		goto error;
+	case DNS_EAI_FAIL:
+		LOG_ERR("DNS resolve failed");
+		goto error;
+	case DNS_EAI_NODATA:
+		LOG_ERR("Cannot resolve address");
+		goto error;
+	case DNS_EAI_ALLDONE:
+		break;
+	case DNS_EAI_INPROGRESS:
+		break;
+	default:
+		LOG_ERR("DNS resolving error");
+		goto error;
+	}
+
+	if (info && info->ai_family == AF_INET) {
+		hr_family = "IPv4";
+		addr = &net_sin(&info->ai_addr)->sin_addr;
+		buffer = net_addr_ntop(info->ai_family, addr, hr_addr, sizeof(hr_addr));
+		memcpy(user_ip, buffer, strlen(buffer));
+	}
+
+error:
+	k_sem_give(&_ctx.semaphore);
+}
+
+static bool start_coap_client(void)
+{
+	static struct sockaddr_in addr = { .sin_family = AF_INET,
+					   .sin_port = htons(5683) };
+	char ip[MAX_IP_SIZE];
+	u16_t dns_id;
+
+	/* Starts the semaphore to wait for the DNS callback */
+	k_sem_init(&_ctx.semaphore, 0, 1);
+
+	memset(&ip, 0, MAX_IP_SIZE);
+	if (dns_get_addr_info(UPDATEHUB_SERVER, DNS_QUERY_TYPE_A, &dns_id,
+			      dns_cb, &ip, NETWORK_TIMEOUT) < 0) {
+		LOG_ERR("Could not resolve dns");
+		return false;
+	}
+
+	k_sem_take(&_ctx.semaphore, NETWORK_TIMEOUT);
+
+	if (net_addr_pton(AF_INET, ip, &addr.sin_addr)) {
+		LOG_ERR("Invalid peer IPv4 address");
+		return false;
+	}
+
+	sock = socket(addr.sin_family, SOCK_DGRAM, IPPROTO_UDP);
+	if (sock < 0) {
+		LOG_ERR("Failed to create UDP socket");
+		return false;
+	}
+
+	if (connect(sock, (struct sockaddr *)&addr,
+		    sizeof(addr)) < 0) {
+		LOG_ERR("Cannot connect to UDP remote");
+		return false;
+	}
+
+	prepare_fds();
+
+	return true;
+}
+
+static void cleanup_conection(void)
+{
+	if (close(sock) < 0) {
+		LOG_ERR("Could not close the socket");
+	}
+
+	memset(&fds[1], 0, sizeof(fds[1]));
+	nfds = 0;
+}
+
+static int send_request(enum coap_msgtype msgtype, enum coap_method method,
+			enum updatehub_uri_path type)
+{
+	struct coap_packet request_packet;
+	u8_t *data = k_malloc(MAX_PAYLOAD_SIZE);
+	int ret = -1;
+	const u8_t *api_header =
```

Comment:
@chtavares592  please define a constant inside our private .h as UPDATEHUB_API_HEADER; this makes the code easier to read.